lib/srv/usermgmt.go

```diff
@@ -0,0 +1,297 @@
+		return nil, nil, trace.Wrap(err)
+	}
+
+	if tempUser != nil {
+		// try to delete even if the user already exists as only users
+		// in the teleport-system group will be deleted and this way
+		// if a user creates multiple sessions the account will
+		// succeed in deletion
+		return tempUser, &userCloser{
+			username: name,
+			users:    u,
+			backend:  u.backend,
+		}, trace.AlreadyExists("User %q already exists", name)
+	}
+
+	groups := append(ui.Groups, types.TeleportServiceGroup)
+	var errs []error
+	for _, group := range groups {
+		if err := u.createGroupIfNotExist(group); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+	}
+	if err := trace.NewAggregate(errs...); err != nil {
+		return nil, nil, trace.WrapWithMessage(err, "error while creating groups")
+	}
+
+	err = u.doWithUserLock(func(_ types.SemaphoreLease) error {
+		if err := u.storage.UpsertHostUserInteractionTime(u.ctx, name, time.Now()); err != nil {
+			return trace.Wrap(err)
+		}
+
+		err = u.backend.CreateUser(name, groups)
+		if err != nil && !trace.IsAlreadyExists(err) {
+			return trace.WrapWithMessage(err, "error while creating user")
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	tempUser, err = u.backend.Lookup(name)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return tempUser, &userCloser{
+		username: name,
+		users:    u,
+		backend:  u.backend,
+	}, nil
+}
+
+func (u *HostUserManagement) doWithUserLock(f func(types.SemaphoreLease) error) error {
+	lock, err := services.AcquireSemaphoreWithRetry(u.ctx,
+		services.AcquireSemaphoreWithRetryConfig{
+			Service: u.storage,
+			Request: types.AcquireSemaphoreRequest{
+				SemaphoreKind: types.SemaphoreKindHostUserModification,
+				SemaphoreName: "host_user_modification",
+				MaxLeases:     1,
+				Expires:       time.Now().Add(time.Second * 20),
+			},
+			Retry: utils.LinearConfig{
+				Step: time.Second * 5,
+				Max:  time.Minute,
+			},
+		})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer u.storage.CancelSemaphoreLease(u.ctx, *lock)
+	return trace.Wrap(f(*lock))
+}
+
+func (u *HostUserManagement) createGroupIfNotExist(group string) error {
+	_, err := u.backend.LookupGroup(group)
+	if err != nil && err != user.UnknownGroupError(group) {
+		return trace.Wrap(err)
+	}
+	err = u.backend.CreateGroup(group)
+	if trace.IsAlreadyExists(err) {
+		return nil
+	}
+	return trace.Wrap(err)
+}
+
+// DeleteAllUsers deletes all host users in the teleport service group.
+func (u *HostUserManagement) DeleteAllUsers() error {
+	users, err := u.backend.GetAllUsers()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	teleportGroup, err := u.backend.LookupGroup(types.TeleportServiceGroup)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = u.doWithUserLock(func(l types.SemaphoreLease) error {
+		var errs []error
+		for _, name := range users {
+			lt, err := u.storage.GetHostUserInteractionTime(u.ctx, name)
```

Comment:
Each user's last interaction time can only ever increase, so we can read it outside of the lock, and only if it's not recent enough grab the lock and check again before deleting the user.